src/main/java/io/cdap/plugin/servicenow/sink/service/ServiceNowSinkAPIRequestImpl.java

```diff
@@ -0,0 +1,197 @@
+          if (servicedRequestsArray.get(i).getAsJsonObject().get("status_code").getAsInt() == HttpStatus.SC_FORBIDDEN) {
+            throw new RuntimeException("Permission denied for " + config.getOperation() + " operation");
+          }
+        }
+        JsonArray unservicedRequestsArray = responseJSON.get(UNSERVICED_REQUESTS).getAsJsonArray();
+        if (unservicedRequestsArray.size() > 0) {
+          if (retryCounter == 1) {
+            throw new RuntimeException("Please decrease the Max Records per Batch while configuring ServiceNow " +
+                                            "Sink Plugin or increase the REST Batch API request timeout property in " +
+                                            "ServiceNow Transaction Quota Rules");
+          }
+          LOG.info("Optimum Max Records per Batch is {}", (servicedRequestsArray.size() - 1));
+          retryUnservicedRequests(records, unservicedRequestsArray);
+        }
+      }
+    } catch (OAuthSystemException | OAuthProblemException | UnsupportedEncodingException | InterruptedException e) {
+      LOG.error("Error in creating a new record", e);
+      throw new RuntimeException("Error in creating a new record");
+    }
+    // Reset request counter
+    counter = 1;
+    // Reset retry counter
+    retryCounter = 0;
+    return apiResponse.getHttpStatus() == HttpStatus.SC_OK;
+  }
+
+  private String getErrorMessage(String responseBody) {
+    try {
+      Gson gson = new Gson();
+      JsonObject jo = gson.fromJson(responseBody, JsonObject.class);
+      return jo.getAsJsonObject("error").get("message").getAsString();
+    } catch (Exception e) {
+      return e.getMessage();
+    }
+  }
+
+  public ServiceNowBatchRequest getPayloadRequest(List<RestRequest> restRequests) {
+    ServiceNowBatchRequest payloadRequest = new ServiceNowBatchRequest();
+    payloadRequest.setBatchRequestId(batchRequestIdCounter.toString());
+    payloadRequest.setRestRequests(restRequests);
+    batchRequestIdCounter++;
+
+    return payloadRequest;
+  }
+
+  private void retryUnservicedRequests(List<RestRequest> records, JsonArray unservicedRequestsArray)
+    throws InterruptedException {
+    List<RestRequest> unservicedRequests = new ArrayList<>();
+    List<Integer> unservicedRequestsIds = new ArrayList();
+    for (int i = 0; i < unservicedRequestsArray.size(); i++) {
+      unservicedRequestsIds.add(unservicedRequestsArray.get(i).getAsInt());
+    }
+    Collections.sort(unservicedRequestsIds);
+    int start = unservicedRequestsIds.get(0);
+    int end = unservicedRequestsIds.get(unservicedRequestsIds.size() - 1);
+    // i = start-1 because request just prior to unserviced request fail due to maximum execution time getting exceeded
+    for (int i = start - 1; i <= end; i++) {
+      unservicedRequests.add(records.get(i - 1));
+    }
+    LOG.info("Retrying unserviced requests from Request No. {} to {}", (start - 1), end);
+    Thread.sleep(1000);
```

Comment:
And how are we setting 1000, any documentation ?

Reply:
Removed it as it is no longer required.